File: macos/flutter_js/Sources/flutter_js/FlutterJsPlugin.swift
import Cocoa
import FlutterMacOS

/// On macOS the JavaScript engine runs entirely in Dart over `dart:ffi` against
/// JavaScriptCore, so this plugin only exists to keep the `io.abner.flutter_js`
/// method channel alive for platform introspection.
public class FlutterJsPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "io.abner.flutter_js", binaryMessenger: registrar.messenger)
    let instance = FlutterJsPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("macOS " + ProcessInfo.processInfo.operatingSystemVersionString)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: android/src/main/kotlin/io/abner/flutter_js/FlutterJsPlugin.kt
package io.abner.flutter_js

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * FlutterJsPlugin
 *
 * On Android the JavaScript engine runs entirely in Dart over `dart:ffi` against the QuickJS
 * code asset built by `hook/build.dart`, so this plugin only exists to keep the
 * `io.abner.flutter_js` method channel alive for platform introspection.
 */
class FlutterJsPlugin : FlutterPlugin, MethodCallHandler {
    private var methodChannel: MethodChannel? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(binding.binaryMessenger, "io.abner.flutter_js").apply {
            setMethodCallHandler(this@FlutterJsPlugin)
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "getPlatformVersion") {
            result.success("Android ${android.os.Build.VERSION.RELEASE}")
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel?.setMethodCallHandler(null)
        methodChannel = null
    }
}
